apps/aechannel/src/aesc_fsm.erl

```diff
@@ -618,81 +638,100 @@ awaiting_signature(cast, {?SIGNED, ?FND_CREATED, SignedTx} = Msg,
+        end, D0);
 awaiting_signature(cast, {?SIGNED, ?UPDATE, SignedTx} = Msg,
-                   #data{latest = {sign, ?UPDATE, OCTx, Updates}} = D) ->
+                   #data{op = #op_sign{tag = ?UPDATE
+                                     , data = OpData0}} = D) ->
+    #op_data{updates = Updates} = OpData0,
     maybe_check_sigs(SignedTx, Updates, aesc_offchain_tx, not_offchain_tx, me,
         fun() ->
             D1 = send_update_msg(
                   SignedTx, Updates,
                   D#data{state = aesc_offchain_state:set_half_signed_tx(
                                     SignedTx, D#data.state)}),
+            OpData = OpData0#op_data{signed_tx = SignedTx},
             next_state(awaiting_update_ack,
-                      log(rcv, ?SIGNED, Msg, D1#data{latest = {ack, ?UPDATE,
-                                                               OCTx, Updates}}))
+                      log(rcv, ?SIGNED, Msg,
+                          D1#data{op = #op_ack{tag = ?UPDATE
+                                             , data = OpData}}))
         end, D);
 awaiting_signature(cast, {?SIGNED, ?UPDATE_ACK, SignedTx} = Msg,
-                   #data{latest = {sign, ?UPDATE_ACK, _OCTx, Updates}, opts = Opts} = D) ->
+                   #data{op = #op_sign{tag = ?UPDATE_ACK
+                                     , data = OpData0}
+                       , opts = Opts} = D) ->
+    #op_data{updates = Updates} = OpData0,
     maybe_check_sigs(SignedTx, Updates, aesc_offchain_tx, not_offchain_tx, both,
         fun() ->
             D1 = send_update_ack_msg(SignedTx, D),
             {OnChainEnv, OnChainTrees} =
                 aetx_env:tx_env_and_trees_from_top(aetx_contract),
             State = aesc_offchain_state:set_signed_tx(SignedTx, Updates, D1#data.state,
                                                       OnChainTrees, OnChainEnv, Opts),
-            D2 = D1#data{log   = log_msg(rcv, ?SIGNED, Msg, D1#data.log),
-                        state = State,
-                        latest = undefined},
+            D2 = D1#data{log   = log_msg(rcv, ?SIGNED, Msg, D1#data.log)
+                       , state = State
+                       , op = ?NO_OP},
             next_state(open, D2)
         end, D);
 awaiting_signature(cast, {?SIGNED, ?SHUTDOWN, SignedTx} = Msg,
-                   #data{latest = {sign, ?SHUTDOWN, _, Updates}} = D) ->
+                   #data{op = #op_sign{tag = ?SHUTDOWN
+                                     , data = OpData0}} = D) ->
+    #op_data{updates = Updates} = OpData0,
     maybe_check_sigs(SignedTx, Updates, aesc_close_mutual_tx, not_close_mutual_tx, me,
         fun() ->
             D1 = send_shutdown_msg(SignedTx, D),
-            D2 = D1#data{latest = {shutdown, SignedTx, Updates},
+            OpData = OpData0#op_data{signed_tx = SignedTx},
+            D2 = D1#data{op = #op_ack{tag = shutdown
+                                    , data = OpData},
```

Comment:
move comma to line below? Also, aren't we doing `{ tag ...` and aligning the comma with the `{`?